auditlog/models/rule.py

```diff
@@ -0,0 +1,398 @@
+                'create', None, new_values)
+            return new_record
+        return create
+
+    def _make_read(self):
+        """Instanciate a read method that log its calls."""
+        # FIXME: read() seems a bit tricky, improve to handle old/new api
+
+        # @api.v7
+        # def read(self, cr, user, ids, fields=None, context=None,
+        #          load='_classic_read', **kwargs):
+        #     print "LOG READ", fields, load, kwargs
+        #     # avoid loops
+        #     if self.env.context.get('auditlog_method_intercepted'):
+        #         return read.origin(
+        #             self, cr, user, ids, fields, context, load, **kwargs)
+        #     # call original method with a modified context
+        #     context = dict(
+        #         self.env.context, auditlog_method_intercepted=True)
+        #     result = read.origin(
+        #         self.with_context(context),
+        #         cr, user, ids, fields, context, load, **kwargs)
+        #     print "RESULT", result
+        #     return result
+
+        # @api.v8
+        # def read(self, fields=None, load='_classic_read', **kwargs):
+        #     print "LOG READ", fields, load, kwargs
+        #     # avoid loops
+        #     if self.env.context.get('auditlog_method_intercepted'):
+        #         return read.origin(self, fields, load, **kwargs)
+        #     # call original method with a modified context
+        #     context = dict(
+        #         self.env.context, auditlog_method_intercepted=True)
+        #     result = read.origin(
+        #         self.with_context(context), fields, load, **kwargs)
+        #     print "RESULT", result
+        #     return result
+
+        def read(self, *args, **kwargs):
+            result = read.origin(self, *args, **kwargs)
+            return result
+        return read
+
+    def _make_write(self):
+        """Instanciate a write method that log its calls."""
+        @api.multi
+        def write(self, vals, **kwargs):
+            rule_model = self.env['auditlog.rule']
+            old_values = dict(
+                (d['id'], d) for d in self.sudo().read(list(self._columns)))
+            result = write.origin(self, vals, **kwargs)
+            new_values = dict(
+                (d['id'], d) for d in self.sudo().read(list(self._columns)))
+            rule_model.sudo().create_logs(
+                self.env.uid, self._name, self.ids,
+                'write', old_values, new_values)
+            return result
+        return write
+
+    def _make_unlink(self):
+        """Instanciate an unlink method that log its calls."""
+        @api.multi
+        def unlink(self, **kwargs):
+            rule_model = self.env['auditlog.rule']
+            rule_model.sudo().create_logs(
+                self.env.uid, self._name, self.ids, 'unlink')
+            return unlink.origin(self, **kwargs)
+        return unlink
+
+    def create_logs(self, uid, res_model, res_ids, method,
```

Comment:
create_logs is more than 70 lines: it's a bit large... Isn't there any useful method you could extract to make overriding easier? (may be not but it's worth thinking about it).